dm/syncer/dml_worker.go

```diff
@@ -112,32 +112,39 @@ func (w *DMLWorker) run() {
 	for i := 0; i < w.workerCount; i++ {
 		queueBucketMapping[i] = queueBucketName(i)
 	}
-
-	for j := range w.inCh {
-		metrics.QueueSizeGauge.WithLabelValues(w.task, "dml_worker_input", w.source).Set(float64(len(w.inCh)))
-		switch j.tp {
-		case flush:
-			w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
-			w.sendJobToAllDmlQueue(j, jobChs, queueBucketMapping)
-			j.flushWg.Wait()
-			w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
-			w.flushCh <- j
-		case asyncFlush:
-			w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
-			w.sendJobToAllDmlQueue(j, jobChs, queueBucketMapping)
-			w.flushCh <- j
-		case conflict:
-			w.addCountFunc(false, adminQueueName, j.tp, 1, j.targetTable)
-			w.sendJobToAllDmlQueue(j, jobChs, queueBucketMapping)
-			j.flushWg.Wait()
-			w.addCountFunc(true, adminQueueName, j.tp, 1, j.targetTable)
-		default:
-			queueBucket := int(utils.GenHashKey(j.dml.key)) % w.workerCount
-			w.addCountFunc(false, queueBucketMapping[queueBucket], j.tp, 1, j.targetTable)
-			startTime := time.Now()
-			w.logger.Debug("queue for key", zap.Int("queue", queueBucket), zap.String("key", j.dml.key))
-			jobChs[queueBucket] <- j
-			metrics.AddJobDurationHistogram.WithLabelValues(j.tp.String(), w.task, queueBucketMapping[queueBucket], w.source).Observe(time.Since(startTime).Seconds())
+	for {
+		select {
+		case <-w.syncCtx.Ctx.Done():
```

Comment:
this will discard jobs in the channel
- dml job will be ignored without setting `s.execError` by fatalFunc which will cause checkpoint not aware about it and forwarded wrongly.
- flush job is not passed into out channel, may cause the generator of flush job waiting forever.

that's critical, please add a test for it.

Reply:
really thx, current handle job logic not allow us do ungraceful close job pipline now, so i revert this change in 
https://github.com/pingcap/tiflow/pull/4035/commits/05ca26babf53aa0de37cd296a27cf31ffd966e25
patl @GMHDBJD

Reply:
LGTM，but I think if ctx close，flush checkpoint should failed immediately

Reply:
seems `w.execError` will make sure FlushPointsExcept will not triggerd when ctx canceled , i am not sure if this is enough

https://github.com/pingcap/tiflow/blob/master/dm/syncer/checkpoint_flush_worker.go#L74